### PR TITLE
[OFFAPPS-992] Bugfixes and refactor

### PR DESCRIPTION
**Pull request by ocke**
/cc @zendesk/vegemite @zendesk/apps-migration 

Bugfixes and complete refactor of some code.

Go by **commits**, for easier readablity

Bugfix: The topbar doesn't open any more when the background receives a message. The 'tests' for checking if you can actually receive a message is now done in the background. That means that top_bar never even gets the message, if it wasn't for the agent.

Bugfix: The page doesn't refresh anymore when senderGroups get updated, if the update doesn't apply to the agent.

The main jest is that I had to move a lot of code, so I had to 'upgrade' the background location to a webpack/babel parsed location.

I also found a problem with the I18n, Handlebars was being include into the new background location, so in order to fix that, I moved the I18n/Handlebars code to the ticket_sidebar location.

Last but no least: I implemented a **caching** mechanism in the client. Before I was doing something like this:

```
function foobar() {
  client.request('users').then((data) => {
    storage('users', data)
  })
}

function foo() {
  const users = storage('users')
  // do stuff with users
}

function bar() {
  const users = storage('users')
  // do stuff with users
}
```

Obviously this is not great, but it was the best I could come up with. However, I made a change now:

```
function foo() {
  client.request('users').then((data) => {
    // doo stuff with users
  })
}

function bar() {
  client.request('users').then((data) => {
    // doo stuff with users
  })
}
```

`client.request` will cache the call internally, so it won't make 2 requests! This makes the code easier to read, more pure, easier to tests, and no slower! win-win-win

Jira: https://zendesk.atlassian.net/browse/OFFAPPS-992

### CODE REVIEW

File: src/javascript/app.js [hunk 1]
```diff
@@ -55,75 +27,118 @@ const appReady = new Promise((resolve) => {
...
-      isSender: isAdmin || app.hasAuthorisedSender(storage('myGroupIds'))
+    if ((event.ctrlKey || event.metaKey) && event.which === 13) {
+      app.sendMsg()
+    }
+  },
+
+  onMessageCloseClick: function (event) {
+    event.preventDefault()
+    const $notification = event.eventTarget.parentNode
+    storage('notifications', storage('notifications').filter((notification) => {
```
Comment (yu-tian113) [inline src/javascript/app.js]: what do you think of using a map instead of array for notifications
{uuid: {message, sender}}

File: src/javascript/lib/i18n.js [hunk 2]
```diff
@@ -60,13 +59,6 @@ const I18n = {
...
 
-Handlebars.registerHelper('t', function (key, context) {
-  try {
-    return I18n.t(key, context.hash)
-  } catch (e) {
-    console.error(e)
-    return e.message
-  }
-})
+I18n.loadTranslations(defaultLocale)
```
[the remark this answers is not in the document]
Comment (ocke) [reply src/javascript/lib/i18n.js]: Good question, sometimes an error occurs *before* we have called `loadTranslations` with the correct locale. So we load the default locale first and then once we have the actual agent's locale, we load those.

[the remark this answers is not in the document]
Comment (ocke) [reply src/javascript/lib/i18n.js]: So technically if we don't do this (which I didn't used to do) you can have an error with 'no such translations...'

File: src/javascript/lib/helpers.js [hunk 3]
```diff
@@ -152,3 +213,44 @@ export function storage (key, value) {
...
 }
+
+// Function to fix arrays that are accidentally an object
+export function fixArray (obj) {
+  return Object.keys(obj).map((key) => {
+    return obj[key]
+  })
+}
+
+export function getClients (arrayOrString) {
```
[the remark this answers is not in the document]
Comment (ocke) [reply src/javascript/lib/helpers.js]: How to deal with multiple locations that are the same?

File: src/javascript/lib/client.js [hunk 4]
```diff
@@ -48,4 +48,30 @@ client.get = function (stringOrArray) {
   })
 }
 
+client.request = function (obj) {
```
Comment (vaevictis) [inline src/javascript/lib/client.js]: Maybe use a more specific name? I find overriding otherwise existing functions and adding side-effects to them while keeping the same name a recipe for confusion. maybe `client.cachedRequest` or `cacheableRequest`?

File: src/javascript/lib/helpers.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (vaevictis) [reply src/javascript/lib/helpers.js]: What's the point of / why do we end up with multiple locations that are the same? Isn't that a smell in the first place?

File: src/javascript/lib/requests.js [hunk 5]
```diff
@@ -30,17 +30,19 @@ export default {
     }
   },
 
-  getGroups: function () {
+  groups: function () {
```
Comment (vaevictis) [inline src/javascript/lib/requests.js]: Whenever I read about caching, I shiver.
We are confident no client will complain some update in group memberships are not propagated to their app that stays opened days at once?
I _believe_ it should be fine in this context, but, well, _cache invalidation is one of the two most difficult things in computer engineering_.

File: src/javascript/app.js [hunk 6]
```diff
@@ -1,4 +1,4 @@
-import { ajax, ajaxPaging, appResize, setting, storage, getClient, anyOverlap } from './lib/helpers'
+import { $, ajax, ajaxPaging, appResize, setting, storage, getClient, anyOverlap, delegateEvent, escapeSpecialChars } from './lib/helpers'
```
Comment (vaevictis) [inline src/javascript/app.js]: As stated in the PR meeting, I think re-using `$` is dangerous as it's so widely associated with `jquery`. Maybe naming it something that communicates _what you're using it for_?

File: src/javascript/lib/datalist.js [hunk 7]
```diff
@@ -148,11 +156,11 @@ export default class Datalist {
...
   filterMenuOptions () {
     const $datalist = this.$datalist()
-    const $input = $datalist.find('.selector input')
-    const text = $input.val().toLowerCase().trim()
+    const $input = $datalist.querySelector('.selector input')
+    const text = ($input.value || '').toLowerCase().trim()
     const regexp = new RegExp(text)
 
-    $datalist.find('.menu_options li').get().forEach((li) => {
+    $datalist.querySelectorAll('.menu_options li').forEach((li) => {
```
[the remark this answers is not in the document]
Comment (ocke) [reply src/javascript/lib/datalist.js]: good find

File: src/javascript/lib/i18n.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (yu-tian113) [reply src/javascript/lib/i18n.js]: 👍

File: src/javascript/lib/helpers.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (ocke) [reply src/javascript/lib/helpers.js]: If you open multiple tickets you will have multiple 'ticket_sidebar' or 'new_ticket_sidebar' locations.

[the remark this answers is not in the document]
Comment (yu-tian113) [reply src/javascript/lib/helpers.js]: 👌

File: src/javascript/lib/requests.js (hunk 5 above)
Comment (ocke) [reply src/javascript/lib/requests.js]: I completely agree. We should be very careful around caching because it sounds easy, but it never is and can be the source of many bugs and/or confusion (around code).

That being said, there are a few things I think we *can* cache. Multiple times I need the groups of a user. Do we need to request that every time? I don't think so, so therefore we can cache that. It's only the simple things I'm willing to cache (users/groups/ticket_fields) but I wouldn't cache tickets/comments/audits.

There is not cache invalidation system (for now), because I didn't need it (for now).

Thoughts?

File: src/javascript/app.js (hunk 6 above)
Comment (ocke) [reply src/javascript/app.js]: how about `query` or `qs` (query selector) or `selector` or `getElement` ?

File: src/javascript/lib/client.js (hunk 4 above)
Comment (ocke) [reply src/javascript/lib/client.js]: I can indeed do this, but I also overwrite `get`. On top of that `client.cachedRequest` doesn't make the request cached. The part that determines if something is cached is in the `request.js` object with the property `cachable: true`. Which might be very confusing for the dev: "I'm using this `cachedRequest` function, but it doesn't cache".

Daniel also pointed out what you are pointing out, and changing the function names sounds a bit strange to me, what we can do is change the client name from `client` to something else to denote we are not using the actual client directly.

Or we can all use the same client wrappers and then it would be clear 😆. I mean the client get/set/invoke/request functions are a bit limited and the reason I wrapped client is to make it easier to work with. For example, I dislike this so much:

```
client.get('ticket.id').then((data) => {
  // data['ticket.id']
})
client.get(['ticket.id', 'ticket.subject']).then((data) => {
  // data['ticket.id']
  // data['ticket.subject']
})
```

First of all, string in, object out, array in, object out. 😠  Also I need to repeat that same damn name. Facebook/aws/google, i looked at each of their sdk's and none of them do something so silly.

```
client.get('ticket.id').then((ticketId) => {
  // ticketId
})
client.get(['ticket.id', 'ticket.subject']).then((data) => {
  const [ id, subject ] = data
})
```

String in, data out, array in, array out (with the same order!). This is also how google/facebook/aws does it. It makes the code so much more readable, especially because I use the first (get with a string) so much more, and I get the data returned I'm looking for.

Anyway, long story short, what should we name client? :P

Comment (vaevictis) [reply src/javascript/lib/client.js]: Ok I agree about changing the client name. Something that describes what's going on, maybe `wrappedClient` or `extendedClient`?

File: src/javascript/lib/helpers.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (vaevictis) [reply src/javascript/lib/helpers.js]: Makes sense. 👌

File: src/javascript/lib/requests.js (hunk 5 above)
Comment (yu-tian113) [reply src/javascript/lib/requests.js]: I can see some potential risk with caching outdated userGroups. But if it's not in the scope that the app needs to handle group changes instantly, I'm ok to leave it.

**Comment by yu-tian113** [comment]
I'm happy with the extended client name

**Comment by vaevictis** [comment]
Me too. Go for gold and :ship::it:
